### PR TITLE
Support for IS NULL condition in cypher and refactor

### PR DESCRIPTION
**Pull request by amitsuryavanshi**
- Refactor to address codeclimate issues
- Added support for 'IS NULL' condition for attribute

### CODE REVIEW

File: cancancan-neo4j.gemspec [hunk 1]
```diff
@@ -22,12 +22,13 @@ Gem::Specification.new do |spec|
...
+  spec.add_development_dependency 'rake'
   spec.add_development_dependency 'rspec', '~> 3.2'
   spec.add_development_dependency 'rubocop', '~> 0.48.1'
   spec.add_development_dependency 'simplecov', '~> 0.12'
   spec.add_development_dependency 'coveralls', '~> 0.8'
   spec.add_development_dependency 'codeclimate-test-reporter', '~> 1.0'
   spec.add_development_dependency 'pry', '~> 0.11.3'
-  spec.add_development_dependency('neo4j-rake_tasks', '>= 0.3.0')
+  spec.add_development_dependency 'neo4j-rake_tasks', '>= 0.3.0'
+  spec.add_development_dependency 'neo4j-community', '~> 2.0'  if RUBY_PLATFORM =~ /java/
```
Comment (amitsuryavanshi) [inline cancancan-neo4j.gemspec]: `neo4j-community` is required to run tests in jruby environment, it is same as neo4j gem.

File: lib/cancancan/neo4j/cypher_constructor_helper.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+          rel_type = relationship_type(relationship)
+          case relationship.direction
+          when :out
+            "-#{rel_type}->"
+          when :in
+            "<-#{rel_type}-"
+          when :both
+            "-#{rel_type}-"
+          end
+        end
```
Comment (klobuczek) [inline lib/cancancan/neo4j/cypher_constructor_helper.rb]: Were you not able to leverage neo4j or neo4j-core for any of this?

**Comment by amitsuryavanshi** [comment]
Hi @klobuczek CC is showing 23 issues, out of which 10 are complexity issues. Code is relatively less complex now, so I think we can merge this. Thanks.

File: spec/cancancan/model_adapters/neo4j_adapter_spec.rb [hunk 3]
```diff
@@ -129,7 +136,7 @@
             it 'fetches all articles with matching conditions' do
               @article2.mentions << Mention.create!
               @ability.cannot :read, Article, mentions: {user: nil}, published: false
-              expect(Article.accessible_by(@ability).to_a).to eq([@cited])
+              expect(Article.accessible_by(@ability).to_a).to eq([@cited, @article2])
```
Comment (amitsuryavanshi) [inline spec/cancancan/model_adapters/neo4j_adapter_spec.rb]: cypher executed here is 
`"MATCH (article:`Article`) WHERE (NOT((article.published=false) AND ( NOT (article:`Article`)-[:`mention`]->()-[:`user`]->())))"` 
and earlier it was
`"MATCH (article:`Article`) WHERE NOT(article.published=false) MATCH (article)-[rel1:`mention`]->(mention:`Mention`) WHERE NOT( NOT (mention)-[:`user`]->())"`

The current one is more accurate given the rule,
`cannot :read, Article, mentions: {user: nil}, published: false`

**Pull request merged**